File: examples/hash/fsharp/func.fs
open System
open System.IO
open System.Security.Cryptography

let createChecksum (data:byte array) =
    use md5 = MD5.Create ()
    let hash = data |> md5.ComputeHash
    BitConverter.ToString(hash).Replace("-", "").ToLower ()

let downloadRemoteImageFile uri =
    async {
        let req = System.Net.HttpWebRequest.Create (System.Uri (uri)) :?> System.Net.HttpWebRequest
        let! res = req.AsyncGetResponse ()
        let stm = res.GetResponseStream ()
        use ms = new MemoryStream ()
        stm.CopyTo(ms)
        return ms.ToArray ()
    }

[<EntryPoint>]
let main argv =
    let isPipedInput =
        try
            Console.KeyAvailable |> ignore
            false
        with
        | _ -> true

    match isPipedInput with
    | true ->
        let input = stdin
        async {
            let! uri = input.ReadToEndAsync () |> Async.AwaitTask
            return! downloadRemoteImageFile uri
        } |> Async.RunSynchronously
        |> createChecksum
        |> printfn "%s"
    | false -> () // if nothing is being piped in, then exit
    0

File: examples/UPDATE-OR-DELETE/hash/csharp/func.cs
using System;
using System.Text;
using System.Security.Cryptography;
using System.IO;

namespace ConsoleApplication
{
    public class Program
    {
        public static void Main(string[] args)
        {
            // if nothing is being piped in, then exit
            if (!IsPipedInput())
                return;

            var input = Console.In.ReadToEnd();
            var stream = DownloadRemoteImageFile(input);
            var hash = CreateChecksum(stream);
            Console.WriteLine(hash);
        }

        private static bool IsPipedInput()
        {
            try
            {
                bool isKey = Console.KeyAvailable;
                return false;
            }
            catch
            {
                return true;
            }
        }
        private static byte[] DownloadRemoteImageFile(string uri)
        {

            var request = System.Net.WebRequest.CreateHttp(uri);
            var response = request.GetResponseAsync().Result;
            var stream = response.GetResponseStream();
            using (MemoryStream ms = new MemoryStream())
            {
                stream.CopyTo(ms);
                return ms.ToArray();
            }
        }
        private static string CreateChecksum(byte[] stream)
        {
            using (var md5 = MD5.Create())
            {
                var hash = md5.ComputeHash(stream);
                var sBuilder = new StringBuilder();

                // Loop through each byte of the hashed data
                // and format each one as a hexadecimal string.
                for (int i = 0; i < hash.Length; i++)
                {
                    sBuilder.Append(hash[i].ToString("x2"));
                }

                // Return the hexadecimal string.
                return sBuilder.ToString();
            }
        }
    }
}
